File: app/schemas/scout.py
"""Phase 15A/B/C/D — Code Scout schemas."""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.schemas.convert import ConvertedLegOut


class ScoutSafetyEdit(BaseModel):
    kind: str
    severity: str = "medium"
    title: str
    detail: str
    leg_index: int | None = None


class ScoutedCodeOut(BaseModel):
    id: int
    code_text: str
    bookmaker: str
    source: str
    source_label: str | None = None
    source_url: str | None = None
    folds: int | None = None
    combined_odds: Decimal | None = None
    risk_band: str = "unknown"
    title: str | None = None
    notes: str | None = None
    scouted_at: datetime | None = None
    hub_url: str | None = None
    # Phase 15B
    verification: str = "unverified"
    confidence_pct: float | None = None
    confidence_label: str | None = "Unverified — copy only."
    legs_count: int = 0
    legs_matched: int = 0
    # Phase 15C
    safety_edits: list[ScoutSafetyEdit] = Field(default_factory=list)
    safety_summary: str | None = None
    # Phase 15D
    convertible: bool = False
    convert_target: str | None = None


class ScoutListResponse(BaseModel):
    count: int
    bookmaker: str
    codes: list[ScoutedCodeOut]
    message: str


class ScoutCodeCreate(BaseModel):
    code_text: str = Field(min_length=3, max_length=64)
    bookmaker: str = Field(default="sportybet", examples=["sportybet", "bet9ja"])
    source: str = Field(default="manual")
    source_label: str | None = None
    source_url: str | None = None
    folds: int | None = Field(default=None, ge=1, le=50)
    combined_odds: Decimal | None = Field(default=None, gt=1)
    title: str | None = None
    notes: str | None = None


class ScoutRefreshResponse(BaseModel):
    status: str
    upserted: int
    sources: list[str]
    message: str


class ScoutConvertResponse(BaseModel):
    convertible: bool
    source_book: str
    target_book: str
    code_text: str
    slip_text: str | None = None
    legs: list[ConvertedLegOut] = Field(default_factory=list)
    matched_count: int = 0
    combined_sportybet: Decimal | None = None
    combined_bet9ja: Decimal | None = None
    combined_best_mixed: Decimal | None = None
    combined_target: Decimal | None = None
    place_summary: str = ""
    message: str

File: mobile/src/api/scout.ts
import { getJson, postJson } from './client';

export type ScoutedCode = {
  id: number;
  code_text: string;
  bookmaker: string;
  source: string;
  source_label?: string | null;
  source_url?: string | null;
  folds?: number | null;
  combined_odds?: number | string | null;
  risk_band: string;
  title?: string | null;
  notes?: string | null;
  scouted_at?: string | null;
  hub_url?: string | null;
};

export type ScoutListResponse = {
  count: number;
  bookmaker: string;
  codes: ScoutedCode[];
  message: string;
};

export type ScoutSort =
  | 'odds_desc'
  | 'odds_asc'
  | 'folds_desc'
  | 'folds_asc'
  | 'date_desc'
  | 'date_asc';

export type ScoutListParams = {
  bookmaker?: string;
  days?: number;
  min_odds?: number;
  max_odds?: number;
  min_folds?: number;
  max_folds?: number;
  risk_band?: string;
  sort?: ScoutSort;
  limit?: number;
  refresh_if_empty?: boolean;
};

export function fetchScoutCodes(params: ScoutListParams = {}) {
  const q = new URLSearchParams();
  q.set('bookmaker', params.bookmaker || 'sportybet');
  q.set('days', String(params.days ?? 14));
  if (params.min_odds != null) q.set('min_odds', String(params.min_odds));
  if (params.max_odds != null) q.set('max_odds', String(params.max_odds));
  if (params.min_folds != null) q.set('min_folds', String(params.min_folds));
  if (params.max_folds != null) q.set('max_folds', String(params.max_folds));
  if (params.risk_band && params.risk_band !== 'all') q.set('risk_band', params.risk_band);
  q.set('sort', params.sort || 'odds_desc');
  q.set('limit', String(params.limit ?? 80));
  if (params.refresh_if_empty === false) q.set('refresh_if_empty', 'false');
  return getJson<ScoutListResponse>(`/scout/codes?${q}`);
}

export function refreshScoutFeed() {
  return postJson<{ status: string; upserted: number; sources: string[]; message: string }>(
    '/scout/refresh',
    {}
  );
}
